File: backend_backup/api/simulation.py
from fastapi import APIRouter
from backend.models.energy import EnergySimulationStep
from backend.services.digital_twin import DigitalTwin
from backend.simulation.simulator import MicrogridSimulator

router = APIRouter(prefix="/api/simulation", tags=["Simulation"])

@router.post("/step")
def run_simulation_step(payload: EnergySimulationStep):
    state = DigitalTwin.evaluate_state(
        station_id=payload.station_id,
        wind_mps=payload.wind_mps,
        lux=payload.lux,
        temp_c=payload.temp_c,
        battery_kwh=payload.battery_kwh,
        gen_kw=payload.gen_kw
    )
    return {"ok": True, "state": state}

@router.get("/scenarios")
def list_scenarios():
    return {"ok": True, "scenarios": MicrogridSimulator.SCENARIOS}

@router.post("/scenario/{scenario_key}")
def run_scenario(scenario_key: str, battery_kwh: float = 85.0):
    res = MicrogridSimulator.run_scenario(scenario_key, battery_kwh)
    return {"ok": True, "simulation": res}

File: frontend_backup/js/simulation.js
/**
 * POLAR ENERGY AI - WHAT-IF STRESS SIMULATION ENGINE
 * Polar Blizzard, Generator Trip, Midnight Sun & Winter Night Scenario Runner
 */

import { apiFetch, showToast } from './app.js';
import { evaluateMicrogrid } from './dashboard.js';

export function initSimulation() {
  bindScenarioButtons();
}

/**
 * Bind scenario cards / buttons
 */
function bindScenarioButtons() {
  const buttons = document.querySelectorAll('[data-scenario]');
  buttons.forEach(btn => {
    btn.addEventListener('click', async () => {
      const scenarioKey = btn.getAttribute('data-scenario');
      await runScenario(scenarioKey);
    });
  });
}

/**
 * Execute scenario against backend engine and sync sliders
 */
export async function runScenario(scenarioKey) {
  try {
    const res = await apiFetch(`/api/simulation/scenario/${scenarioKey}?battery_kwh=65.0`, {
      method: 'POST'
    });

    if (res && res.ok && res.simulation) {
      const sim = res.simulation;
      const weather = sim.scenario_applied;

      // Update Dashboard range sliders if on dashboard
      setSliderVal('rngWind', 'txtWind', weather.wind_mps, ' m/s');
      setSliderVal('rngLux', 'txtLux', weather.lux, ' Lux');
      setSliderVal('rngTemp', 'txtTemp', weather.temp_c, ' °C');
      if (weather.gen_kw !== undefined) {
        setSliderVal('rngGen', 'txtGen', weather.gen_kw, ' kW');
      }

      showToast(`SIMULATION ACTIVE: ${sim.scenario_name}`, sim.state.is_critical ? 'crit' : 'info');

      // Re-evaluate Digital Twin
      if (typeof evaluateMicrogrid === 'function') {
        evaluateMicrogrid();
      }
    }
  } catch (err) {
    showToast(`Simulation scenario "${scenarioKey}" failed`, 'crit');
  }
}

function setSliderVal(sliderId, textId, value, unit) {
  const el = document.getElementById(sliderId);
  const textEl = document.getElementById(textId);
  if (el) {
    el.value = value;
    // Dispatch input event to sync internal telemetry state
    el.dispatchEvent(new Event('input'));
  }
  if (textEl) {
    textEl.textContent = `${value.toFixed(1)}${unit}`;
  }
}

document.addEventListener('DOMContentLoaded', () => {
  initSimulation();
});

